File: Leetcode/ValidPalindrome.cs
// A phrase is a palindrome if, after converting all uppercase letters into lowercase letters and removing all non-alphanumeric characters, it reads the same forward and backward. Alphanumeric characters include letters and numbers.
// Given a string s, return true if it is a palindrome, or false otherwise.

public class Solution 
{
    public bool IsPalindrome(string s) 
    {
        int left = 0, right = s.Length - 1;
        
        while (left <= right)
        {
            while (!Char.IsLetterOrDigit(s[left]) && left < right)
                left++;
            while (!Char.IsLetterOrDigit(s[right]) && left < right)
                right--;
            
            if (s[left] == s[right] || 
                Char.ToUpper(s[left]) == Char.ToUpper(s[right]))
            {
                left++;
                right--;
            }
            else
                return false;
        }
        return true;
    }
}

File: Leetcode/ValidPalindrome.py
# A phrase is a palindrome if, after converting all uppercase letters into lowercase letters and removing all non-alphanumeric characters, it reads the same forward and backward. Alphanumeric characters include letters and numbers.
# Given a string s, return true if it is a palindrome, or false otherwise.

class Solution(object):
    def isPalindrome(self, s):
        left = 0
        right = len(s) - 1
        
        while left <= right:
            while not s[left].isalnum() and left < right:
                left += 1
            while not s[right].isalnum() and left < right:
                right -= 1
            
            if s[left] == s[right] or s[left].upper() == s[right].upper():
                left, right = left + 1, right - 1
            else:
                return False
        return True